horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/SchemaServiceImpl.java

```diff
@@ -146,29 +157,65 @@ public int idByUri(String uri) {
     @Transactional
     @Override
     public Integer add(Schema schemaDTO) {
-        verifyNewSchema(schemaDTO);
+        // we are creating a new schema, ensure all ids are cleaned up
+        schemaDTO.clearIds();
+
+        log.debugf("Creating new schema: %s", schemaDTO.toString());
+        return addOrUpdateSchema(schemaDTO);
+    }
+
+    @RolesAllowed(Roles.TESTER)
+    @WithRoles
+    @Transactional
+    @Override
+    public Integer update(Schema schemaDTO) {
+        // check the schema already exists in the db
+        if (schemaDTO.id == null || SchemaDAO.findById(schemaDTO.id) == null) {
+            throw ServiceException.notFound("Missing schema id or schema with id " + schemaDTO.id + " does not exist");
+        }
+
+        log.debugf("Updating schema (%d): %s", schemaDTO.id, schemaDTO.toString());
+        return addOrUpdateSchema(schemaDTO);
+    }
+
+    /**
+     * Create or update a Schema entity by performing the appropriate validation logics
+     * @param schemaDTO the DTO that should be mapped to the database
+     * @return the db Schema id
+     */
+    private int addOrUpdateSchema(Schema schemaDTO) {
+        // whether we should trigger a schema synchronization
+        // this happens when creating new schemas or, updating uri or JSON schema
+        boolean syncSchemas = false;
+
+        validateSchema(schemaDTO);
+
         // Note: isEmpty is true for all non-object and non-array nodes
         if (schemaDTO.schema != null && schemaDTO.schema.isEmpty()) {
             schemaDTO.schema = null;
         }
+
         SchemaDAO schema = SchemaMapper.to(schemaDTO);
         if (schemaDTO.id != null && schemaDTO.id > 0) {
-            SchemaDAO existing = SchemaDAO.findById(schema.id);
-            if (existing == null)
-                throw ServiceException.badRequest("An id was given, but it does not exist.");
+            // update existing Schema
+            SchemaDAO existing = (SchemaDAO) SchemaDAO.findByIdOptional(schema.id)
+                    .orElseThrow(() -> ServiceException.badRequest("Cannot find schema with id " + schemaDTO.id));
```

Comment:
Nice, not seen findByIdOptional before, cool!

Reply:
yeah that is quite useful in such cases where you want to throw a custom exception (or put a default) when there is no result